src/v/coproc/script_context.cc

```diff
@@ -299,36 +302,96 @@ ss::future<storage::log> get_log(storage::api& api, const model::ntp& ntp) {
       storage::ntp_config(ntp, api.log_mgr().config().base_dir));
 }
 
-ss::future<bool>
-write_checked(storage::log log, model::record_batch_reader reader) {
+/// Solution to case where scripts writing to the same materialized topic may
+/// attempt to write a record with a lower term_id then the logs base.
+class term_id_barrier {
+public:
+    explicit term_id_barrier(model::term_id last)
+      : _last(last) {}
+
+    ss::future<ss::stop_iteration> operator()(const model::record_batch& rb) {
+        /// If the situation is encountered, the consumer will be alerted,
+        /// and in the case below, the reference_window_consumer will not
+        /// attempt to further process the batch, i.e. aborting the write
+        if (rb.term() < _last) {
+            _exited_early = true;
+        }
+        return ss::make_ready_future<ss::stop_iteration>(
+          _exited_early ? ss::stop_iteration::yes : ss::stop_iteration::no);
+    }
+
+    std::optional<model::term_id> end_of_stream() {
+        return _exited_early ? std::nullopt : std::optional(_last);
+    }
+
+private:
+    bool _exited_early{false};
+    model::term_id _last;
+};
+
+ss::future<std::variant<script_context::write_response, model::term_id>>
+script_context::write_checked(
+  storage::log log,
+  model::term_id last_term,
+  model::record_batch_reader reader) {
+    using rt_val = std::variant<write_response, model::term_id>;
     const storage::log_append_config write_cfg{
       .should_fsync = storage::log_append_config::fsync::no,
       .io_priority = ss::default_priority_class(),
       .timeout = model::no_timeout};
     return std::move(reader)
       .for_each_ref(
         coproc::reference_window_consumer(
-          model::record_batch_crc_checker(), log.make_appender(write_cfg)),
+          model::record_batch_crc_checker(),
+          term_id_barrier(last_term),
+          log.make_appender(write_cfg)),
         model::no_timeout)
-      .then([](std::tuple<bool, ss::future<storage::append_result>> t) {
-          const auto& [crc_parse_success, _] = t;
-          return !crc_parse_success;
+      .then([](auto rs) mutable {
```

Comment:
nit: I thought that we're avoiding this

Reply:
Agreed, I used my judgement. Argument being that the type is pretty long and ugly. I could make a proper typedef I suppose

Reply:
let's do that. 

we should revisit this w/ clang12 but up to clang 11 this bloated the IC (instruction cache)

Reply:
Ok ! Done